### DISCUSSION TITLE
Types in plugins

### DISCUSSION THREAD

**Discussion by maxbrieiev** [vuejs/pinia#578] [Help and Questions] [answered]
Suppose I have a couple of types and store definitions:

```typescript
type AuthState = { accessToken: string; }
type UserState = { name: string }
type States = {
  auth: AuthState;
  user: UserState;
}

const useAuth = defineStore({ id: 'auth' });
const useUser = defineStore({ id: 'user' });
```

Initial state of the stores is populated by plugin. Here is the plugin:

```typescript
const localStoragePlugin = ({ store }: PiniaPluginContext) => {
  const serialized = localStorage.getItem(store.$id);
  if (serialized !== null) {
    const state = <States[typeof store.$id]>JSON.parse(serialized); // <<<------ problem here
    store.$state = state;
    return state;
  }
};
```
I want the types to be narrowed based on the `store.$id`, but the problem is `typeof store.$id` is inferred as `string`. Is it possible to infer it as a literal, that is when we call `useAuth()` it is inferred as `'auth'` and when we call `useUser()` it is inferred as `'user'`?

Thanks.

**Comment by posva** [comment, marked as answer]
I don't think you will be able to infer that because PiniaPluginContext is not aware of defined stores and making it aware would be a very inconvenient manual step anyway

**Reply by maxbrieiev** [reply]
I see, thank you.
